nucleus/common/glassfish-api/src/main/java/org/glassfish/api/invocation/InvocationManagerImpl.java

```diff
@@ -68,347 +71,334 @@
 
-            if (parentInvocation.getInvocationType() == SERVLET_INVOCATION) {
+            ComponentInvocationType parentType = parentFrame.getInvocationType();
+            if (parentType == SERVLET_INVOCATION) {
 
                 // If this is a thread created by user in servlet's service method
                 // create a new ComponentInvocation with transaction
                 // left to null and instance left to null
                 // so that the resource won't be enlisted or registered
 
                 ComponentInvocation invocation = new ComponentInvocation();
-                invocation.setComponentInvocationType(parentInvocation.getInvocationType());
-                invocation.setComponentId(parentInvocation.getComponentId());
-                invocation.setAppName(parentInvocation.getAppName());
-                invocation.setModuleName(parentInvocation.getModuleName());
-                invocation.setContainer(parentInvocation.getContainer());
-                invocation.setJndiEnvironment(parentInvocation.getJndiEnvironment());
-
-                childInvocationArray.add(invocation);
-            } else if (parentInvocation.getInvocationType() != EJB_INVOCATION) {
-
+                invocation.setComponentInvocationType(parentType);
+                invocation.setComponentId(parentFrame.getComponentId());
+                invocation.setAppName(parentFrame.getAppName());
+                invocation.setModuleName(parentFrame.getModuleName());
+                invocation.setContainer(parentFrame.getContainer());
+                invocation.setJndiEnvironment(parentFrame.getJndiEnvironment());
+
+                childFrames.add(invocation);
+            } else if (parentType != EJB_INVOCATION) {
                 // Push a copy of invocation onto the new result
-                // ArrayList
-                ComponentInvocation cpy = new ComponentInvocation();
-                cpy.componentId = parentInvocation.getComponentId();
-                cpy.setComponentInvocationType(parentInvocation.getInvocationType());
-                cpy.instance = parentInvocation.getInstance();
-                cpy.container = parentInvocation.getContainerContext();
-                cpy.transaction = parentInvocation.getTransaction();
-
-                childInvocationArray.add(cpy);
+                childFrames.add(new ComponentInvocation(
+                        parentFrame.getComponentId(),
+                        parentType,
+                        parentFrame.getInstance(),
+                        parentFrame.getContainerContext(),
+                        parentFrame.getTransaction()));
             }
         }
 
-        return childInvocationArray;
+        return childFrames;
     }
 
 
-    static class InvocationArray<T extends ComponentInvocation> extends ArrayList<T> {
+    static final class InvocationFrames extends ConcurrentLinkedDeque<ComponentInvocation> {
```

Comment:
> I'm also afraid that extending rather than delegating makes too many public methods available.

`InvocationFrames` is only used privately. 

Again, done this for peace of mind. Could go to `ArrayDequeue`.